### ISSUE TITLE
Rich text page links always rendered as absolute URLs

### ISSUE THREAD

**Issue by chosak** [wagtail/wagtail#5876]
### Issue Summary

Page links in rich text are always rendered as absolute URLs if there are multiple Wagtail Sites defined, even in cases where they could and should be rendered as relative URLs. A page with a rich text link to another page on the same domain should always use a relative URL; this happens if there's only a single Site defined, but not when multiple are.

This happens generally because rendering of rich text has no context beyond the string being parsed, and more specifically because the request object isn't available to help scope the URL. The [existing link handler logic](https://github.com/wagtail/wagtail/blob/26e2e5341a43df8db451e5d4fe01e80adec7556f/wagtail/core/rich_text/pages.py#L22) uses the `Page.url` property instead of the [recommended](https://docs.wagtail.io/en/v2.8/releases/1.11.html#caching-of-site-level-url-information-throughout-the-request-cycle) `Page.get_url()` method, which, if it had a request object, could generate a different URL depending on which site it's on. 

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. `cd myproject`
3. Edit the HomePage model in home/models.py to add a RichTextField:
   
   ```py
   from django.db import models

   from wagtail.admin.edit_handlers import FieldPanel
   from wagtail.core.fields import RichTextField
   from wagtail.core.models import Page


   class HomePage(Page):
       body = RichTextField(blank=True)

       content_panels = Page.content_panels + [
           FieldPanel('body', classname="full"),
       ]
   ```

4. Edit the HomePage template in home/templates/home/home_page.html to expose the new field:
   
   ```html
   {% extends "base.html" %}

   {% load wagtailcore_tags %}

   {% block content %}
   {{ page.body | richtext }}
   {% endblock content %}
   ```

5. `./manage.py makemigrations && ./manage.py migrate`
6. `./manage.py createsuperuser`
7. `./manage.py runserver`
8. [Edit the homepage](http://localhost:8000/admin/pages/3/edit/) and add an internal link in the body field to itself. Save a draft.
9. [Preview the page](http://localhost:8000/admin/pages/3/edit/preview/), and observe that the link tag is rendered as **`<a href="/">Home</a>`**.
10. [Add a new Site](http://localhost:8000/admin/sites/new/). You can create a new page for its root, or use the same home page (it doesn't matter for purposes of demonstrating this issue).
11. [Preview the page again](http://localhost:8000/admin/pages/3/edit/preview/), and observe that the link tag is now rendered as **`<a href="http://localhost/">Home</a>`**.

I consider this a bug because users should expect that rich text uses Wagtail's ability to generate relative links when possible. Additionally, without a request object, the URL logic can't make optimal use of the [logic around caching site root paths](https://github.com/wagtail/wagtail/blob/26e2e5341a43df8db451e5d4fe01e80adec7556f/wagtail/core/models.py#L782-L793).

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

* Python version: 3.6.9
* Django version: 2.2.8
* Wagtail version: 2.7

**Comment by gasman** [comment]
As far as I can see, this isn't fixable without changing the `|richtext` filter to some other mechanism (e.g. a template tag), as template filters don't have access to any template context beyond the specific variable being passed.

**Comment by Pithikos** [comment]
Yes, this seems like a major defect. Mainly since it means that it's impossible to test a site locally with tools like Cypress.

Is there any workaround?

**Comment by Pithikos** [comment]
So I managed to solve this by overriding the default `richtext` filter with a customized one that can take the current page as argument. Source code [here](https://gist.github.com/Pithikos/f2e70977074fec28e64359ac2f5836a0).

Usage in template
```
<div class="container">
  {{ self.body|richtext:self }}
<div>
```
